File: JavaLearn/src/com/leetcode/longestPalindrome.java
package com.leetcode;

import java.util.Map;

public class longestPalindrome {
    public int longestPalindrome(String s){
        int[] count = new int[128];
        int n = s.length();
        for (int i = 0; i < n; i++){
            char temp = s.charAt(i);
            count[temp]++;
        }
        
        int ans = 0;
        for (int i : count) {
            ans+=i/2*2;
            if (i%2 == 1 && ans % 2 ==0){
                ans++;
            }
        }
        return ans;
    }
}

File: scala_study/src/main/scala/leetcode/longestPalindrome.scala
package leetcode

object longestPalindrome {
  def longestPalindrome(s:String):String ={
    val n = s.length
    val dp = new Array[Array[Int]](n)
    for(i <- 0 until n){
      dp(i) = new Array[Int](n)
    }
    var ans: String = new String

    for (i <- 0 until n){
      for (j <- 0 until n if j+i<n){
        val t = i + j
        if (i == 0) {
          dp(j)(t) = 1
        }else if( i == 1) {
          if (s.charAt(j)==s.charAt(t)){
            dp(j)(t) = 1
          }
        }else{
          if(s.charAt(j)==s.charAt(t)){
            dp(j)(t) = dp(j+1)(t-1)
          }
        }

        if(dp(j)(t) > 0 && i+1 > ans.length){
          ans = s.substring(j,t+1)
        }

      }
    }
    ans
  }

}
